lib/ace/ext/papeeria/tex_completer.coffee

```diff
@@ -86,37 +93,68 @@ define( (require, exports, module) ->
     meta: 'equation'
   )
   
-  init = (editor, bindKey) ->
-    HashHandler = require("ace/keyboard/hash_handler").HashHandler; 
-    keyboardHandler = new HashHandler();
+  # Specific for token's system of type in ace
+  # We saw such a realization in html_completions.js
+  istype = (token, type) -> 
+    return token.type.lastIndexOf(type) > -1
+  
+
+  init = (editor, bindKey) -> 
+    HashHandler = require("ace/keyboard/hash_handler").HashHandler
+    keyboardHandler = new HashHandler()
     keyboardHandler.addCommand(
-      name: 'add item in list mode',
-      bindKey: bindKey,
-      exec: (editor) ->
+      name: 'add item in list mode'
+      bindKey: bindKey
+      exec: (editor) ->  
         pos = editor.getCursorPosition()
-        curLine = editor.session.getLine(pos.row);
+        curLine = editor.session.getLine(pos.row)
+        indentCount = ContextHelper.getNestedListDepth(editor.session, pos.row)
+        tabString = editor.getSession().getTabString()
         # it's temporary fix bug with added \item before \begin{itemize|enumerate}
-        if (editor.getSession().getContext(pos.row) == "list" && curLine.indexOf("begin") < pos.column) 
-          editor.insert("\n\t\\item ")
+        if ContextHelper.getContext(editor.session, pos.row) == LIST_STATE && curLine.indexOf("begin") < pos.column
+          editor.insert("\n" + tabString.repeat(indentCount) + "\\item ")
           return true
         else 
-          return false
+          return false   
     )
     editor.keyBinding.addKeyboardHandler(keyboardHandler)
 
-  getCompletions = (editor, session, pos, prefix, callback) ->
-    context = session.getContext(pos.row)
-    if context == "start"
-      callback(null, listSnippets.concat(equationSnippets.concat(basicSnippets)))
+  class ReferenceGetter
+    constructor: ->
+      @cachedURL =  ""
+      @cache = []
+    processData: (data) -> 
+      @cache = data.map((elem) => 
+          return {
+            name: elem.caption
+            value: elem.caption
+            score: Number.MAX_VALUE
+            meta: "ref"
+          }
+    )
 
-    if context == LIST_STATE
-      callback(null, listKeywords_.concat(listSnippets.concat(equationSnippets)))
+    getReferences: (url, callback) -> 
+      if (url != @cashedURL)
```

Comment:
Hello undefined reference s/cashed/cached

Or even better, rename it. You don't need to cache url, and the point of this variable is not caching. It is just an id of the last received references. maybe `lastFetchedUrl` is ok?